### ISSUE TITLE
Custom Split

### ISSUE THREAD

**Issue by nickolasrm** [kedro-partitioned/kedro-partitioned#3]
Hello,

Is it possible to add a feature in order to add custom split fonctionality ? I want to create a split for a dataset returning list if possible

Thanks !

@adrienpl moved the issue to this repository

**Comment by nickolasrm** [comment]
Sure! I think it is doable. Multinode works like this:

1. A load balancer node receives the full partitioned dataset and splits load between the desired number of slices stores the load splitting result into a json inside the partitioned output
2. Each slice node reads the json, selects its corresponding slice of the original partitioned and processes it saving the result inside the output partitioned dataset
3. A barrier node at the end of the slices section is then added to prevent synchronization problems

I think the feature you are asking would be in the 1st and 2nd steps. Regarding the topics we talked about in our meeting, would you be responsible for adding this feature? If so, I would highly appreciate your contribution. Feel free to send me pull requests, I will review them ASAP.

Would you mind providing an example of use? So that we can discuss load balancing problems as well. Lists are good, but their problem is guaranteeing they can be balanced across nodes correctly. With dict I split the traffic by saving the keys each slice is going to process inside the JSON. However, with lists you would not have a deterministic key to do the splitting. We can use indexes, but we would have to guarantee the items will always stay in the same indexes.

To summarize, I think we should think about the interface and data contracts custom splitting functions should follow in order to allow deterministic load balancing and avoid concurrency problems. For example, maybe custom splitting should allow any dataset to be used, but always return keys for load splitting and a way for reading them in step 2. Does that makes sense?

> **Note**: I'm not disconsidering the feature, but I believe you may also fix the problem by creating a custom dataset that behaves like a partitioned dataset (dict) inheriting the dataset you are using as well

@adrienpl
